api/src/middleware/authorization-middleware.ts

```diff
@@ -0,0 +1,80 @@
+import { type NextFunction, type Response } from "express"
+import { type Request as JwtRequest } from "express-jwt"
+import { isNil } from "lodash"
+
+import logger from "@/utils/logger"
+import { User } from "@/models"
+
+import auth0Integration, { Auth0PayloadError } from "@/integrations/auth0-integration"
+
+import sequelize from "@/models"
+
+export type AuthorizationRequest = JwtRequest & {
+  user?: User
+}
+
+async function findUserFromAuth0Subject(auth0Subject: string): Promise<User | null> {
+  return User.findOne({ where: { sub: auth0Subject } })
+}
+
+export async function ensureUserFromAuth0Token(token: string): Promise<User> {
+  const { auth0Subject, email, firstName, lastName } = await auth0Integration.getUserInfo(token)
+  const user = await findUserFromAuth0Subject(auth0Subject)
+
+  if (!isNil(user)) {
+    return user
+  }
+
+  const newUser = await sequelize.transaction(async (transaction) => {
+    const existingUser = await User.findOne({
+      where: { sub: auth0Subject },
+      transaction,
+    });
+
+    if (existingUser) {
+      return existingUser;
+    }
+    const newUser = await User.create({
+      sub: auth0Subject,
+      email: email,
+      firstName: firstName,
+      lastName: lastName,
+      roles: [User.Roles.USER],
+      status: User.Statuses.ACTIVE,
+    }, { transaction })
+  
+    logger.info(`CREATED USER FOR ${email}: ${JSON.stringify(newUser.dataValues)}`)
+    return newUser
+  })
+
+  return newUser
+}
+
+export async function authorizationMiddleware(
```

Comment:
It might make sense to make this the first function in the file?
For context, JS function's hoist scope so
```
function foo() {
   bar()
}

function bar() {
}
```
is valid.